Palindrome product article (#2104)
exercism/csharp: exercises/practice/palindrome-products/.articles/is-palindrome-check/content.md

> # Is number a palindrome? 

Reviewer: I think I would make this more matter-of-factly, not a question:

```suggestion
# Check if a number is a palindrome
```

There are multiple approaches to solve the palindrome product exercise. 

> Most of them require us to check if a given number is a palindrome. 

Reviewer: Don't all of them require this?

Author: Not all of them. It is possible to generate palindrome numbers in which case you don't need to test for the palindrome. Once you generate the numbers (by composing digits following set patterns) you only have to discover the factors.

Reviewer: Ooh neat!

Author: This could be an interesting follow up article. The solution becomes more complex than the problem itself so I don't think including a generative approach in the approaches will be helpful, but it might be an interesting thing for those who are into digging deep. 

It would create a nice series of challenges - something I proposed in forum but is not really the way exercises work right now. You could have three connected exercises: Check if a number is a palindrome, palindrome product and refactoring to a generative approach. Each next one could have previous code as a starting point. It's a longer discussion for another day

Reviewer: > This could be an interesting follow up article. The solution becomes more complex than the problem itself so I don't think including a generative approach in the approaches will be helpful, but it might be an interesting thing for those who are into digging deep.

Well, it might be actuall! For the reverse string exercise, I also included a Span-based approach which you normally wouldn't use, but could still be interesting to students. I think this would fit well too. But an article would be fine too.

> ## Approaches

Reviewer: I'd remove the header (this is an article, and approaches might be confusing here).

```suggestion
```

Reviewer: Once this is removed, the other headings can become level two headings.

There are multiple ways to do it in C# and we will look at them in this article. 

### String reversal

> Possibly the most obvious solution is to treat a number as a string of characters
> and reverse it. Let's have a look:

Reviewer: You're currently breaking up lines at a cutoff point, but our [style guide](https://exercism.org/docs/building/markdown/markdown#h-one-sentence-per-line) suggests using one sentence per line (no cutoff). That would make this:

```suggestion
Possibly the most obvious solution is to treat a number as a string of characters and reverse it.
Let's have a look:
```

Of course, this would also apply elsewhere.

Reviewer: And maybe be explicit about the last part of the logic too (comparing the strings):

```suggestion
The most straightforward solution is to treat a number as a string of characters and see if it is equal to the reversed string.
Let's have a look:
```

Feel free to change the wording.

```csharp
using System.Linq;

private static bool IsPalindrome(int number)
{
    var original = number.ToString();
    var reversed = new String(original.Reverse().ToArray());
    return original.Equals(reversed);
}
```

If you come from some other languages you might be surprised by the above code.
Unfortunately, C# doesn't come with a built in method to reverse a string. 
We have to use `.Reverse()` extension method provided by 

> [Language-Integrated Query (LINQ)](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/). However, it returns an iterator and so it has to be collected with `.ToArray()` 

Reviewer: Our [markdown guidelines](https://exercism.org/docs/building/markdown/markdown#h-links) suggest to use reference links.

Author: Done

and converted into a `String` with `new String(char[])` constructor. 

> If this looks like a lot of code to you, it might be slightly shortened using 

Reviewer: I would probably put this in a sub-heading.

[string interpolation](https://learn.microsoft.com/en-us/dotnet/csharp/tutorials/string-interpolation)
to convert a number into a text like in the example below. However, before you use it
read the performance considerations below.

```csharp
using System.Linq;

> private static bool IsPalindrome(int number)
> =>  $"{number}".Equals(new String($"{number}".Reverse().ToArray()));

Reviewer: This is more consistent with existing docs:

```suggestion
private static bool IsPalindrome(int number) =>
    $"{number}".Equals(new String($"{number}".Reverse().ToArray()));
```

Author: I had a quick look at recent style guides at [Microsoft Learn](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions), [dotnet project](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md) and [Google](https://google.github.io/styleguide/csharp-style.html). Non of them provide clear guidance what to do in situations where `=>` operator is used to define a function where function has a single statement, but is too long for a single line. The common practice is to run with too long lines which make readability suffer. 

The style which is more common, as you suggest, in the documentation is to leave the `=>` hanging at the end of the line, but then it might be difficult to spot and quickly understand the meaning of the statement. The alternative is to start the next line with the operator. This is something I have recently picked up while mentoring the C# track. Some people out there use it to great effect. I also like how it falls in line with how we see the text. If you haven't seen it I do recommend [Seven Ineffective Coding Habits of Many Programmers by Kevlin Henney](https://www.youtube.com/watch?v=ZsHMHukIlJY), especially [the part about style](https://www.youtube.com/watch?v=ZsHMHukIlJY&t=1027s).

So, with some vagueness around the official style guides and emerging practice, perhaps we could improve here and over time help this become the majority option?

```

### List of digits

Another approach is to use [modulo division](https://en.wikipedia.org/wiki/Modulo)
to extract individual digits from the least significant to the most significant. 
We can capture them in a list, and then read them again using maths again, 
to reconstruct the number in the reverse order. Before we look at the code, 
let's look at an example. The number is 123. 

| Step | number | n % 10 | n / 10 |    list |
|-----:|-------:|-------:|-------:|:--------|
|    1 |    123 |      3 |     12 | [3]     |
|    2 |     12 |      2 |      1 | [3,2]   |
|    3 |      1 |      1 |      0 | [3,2,1] |

So now we have digits, already in reversed order `[3,2,1]`. We have to combine them into a number.

| Step | number | digit | number * 10 + digit |
|-----:|-------:|------:|:--------------------|
|    1 |      0 |     3 | 3                   |
|    2 |      3 |     2 | 32                  |
|    3 |     32 |     1 | 321                 |

And now the same in code:

```csharp
using System.Collections.Generic;

> private static bool IsPalindrome(int number) { 

Reviewer: ```suggestion
private static bool IsPalindrome(int number)
{ 
```

    var original = number;
    var digits = new List<int>();
    

>     while (number > 0) { 

Reviewer: For consistency with other code:

```suggestion
    while (number > 0)
    { 
```

Author: Done

        digits.Add(number % 10);
        number /= 10;
    }

    var reversed = 0;

>     foreach(var digit in digits) {

Reviewer: ```suggestion
    foreach(var digit in digits)
    {
```

Author: Done

        reversed = reversed * 10 + digit;
    }
    
    return original == reversed;
}
```

> ### The math without the list

Reviewer: Maybe just:

```suggestion
### Math (without list)
```

Author: I renamed the headers. Tell me what you think about the latest version.

The list in the above code helps us to see what is going on. 
The numbers are extracted, stored in the list and then collected. 
But, this can be done on the fly, using just math and loops without the storage. 

```csharp
private static bool IsPalindrome(int number) { 
    var original = number;
    var reversed = 0;
    

>     while (number > 0) { 

Reviewer: ```suggestion
    while (number > 0)
    { 
```

        reversed = reversed * 10 + number % 10;
        number /= 10;
    }
    
    return original == reversed;
}
```

> ## The extension method options

Reviewer: ```suggestion
## Converting to an extension method
```

Author: Done

All of the above solutions correctly check if a number is a palindrome. 
They differ in performance, something we will look at in a moment.
But they also differ in readability. You will have to decide which one makes

> the most sense to you. 

Reviewer: I'd probably remove this section.

Author: agreed. reading it again it doesn't add much substance to the content

Talking about readability of code, we have another decision to make. 
All of the above solutions can be implemented like they are above, or 
as an [[extension method](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/extension-methods)]

What does it change? The code reads differently. Instead of 

```csharp
if (IsPalindrome(number)) { /* do something */ }
```

you could have

```csharp
if (number.IsPalindrome()) { /* do something */ }
```

You decide which one fits better with your code and expectations. 

## Performance considerations

> OK, let's talk performance. I have tested the four methods above using

Reviewer: We usually use "we" instead of "I" in our documents.

[BenchmarkDotNet](https://benchmarkdotnet.org/). I won't share the details

> of where I run it, on what hardware as we are interested in relative performance only. 

Reviewer: Could you maybe include the code that was used to run the benchmarks? Like https://github.com/exercism/csharp/tree/main/exercises/practice/leap/.articles/performance/code does. The article itself can then link to the code (see https://exercism.org/tracks/csharp/exercises/leap/articles/performance), which allows students to reproduce it locally.

What is important, that in an individual test I am checking all number between 1 and 100,000

> to see if they are a palindrome or not. Here are the results:

Reviewer: I think this (what numbers are being checked) are a key detail. Ideally, the benchmark would test with different numbers, from small to large. See https://github.com/exercism/csharp/blob/main/exercises/practice/reverse-string/.articles/performance/code/Program.cs#L10-L11 for an example.

Author: I think this could be an interesting aspect but the article is getting long and complex as it is. My intention was to show that there are differences in performance for those curious about things like that, but there isn't much in it. Perhaps we could write a separate article to deal with performance specifically and there it could be explored in much more detail with various options shown. 

talking about performance, is there a way to write a tack level article? I think it would be interesting to explore either tools, or specifically BenchmarkDotNet?

|        Method |        Mean |     Error |    StdDev |
|-------------- |------------:|----------:|----------:|

> |  String1Check | 13,131.6 μs | 100.98 μs |  94.45 μs |
> |  String2Check | 19,578.5 μs | 324.35 μs | 303.40 μs |
> |     ListCheck |  5,690.9 μs |  59.71 μs |  52.93 μs |
> | MathLoopCheck |    935.2 μs |  12.60 μs |  11.79 μs |

Reviewer: Maybe use different method names that use similar names to the headings used in this file?
Something like:

```suggestion
|  ReversedString | 13,131.6 μs | 100.98 μs |  94.45 μs |
|  ReversedStringInterpolated | 19,578.5 μs | 324.35 μs | 303.40 μs |
|     ListOfDigits |  5,690.9 μs |  59.71 μs |  52.93 μs |
| MathLoop |    935.2 μs |  12.60 μs |  11.79 μs |
```

There appears to be a big difference between almost 20ms and just under 1ms. 
It is also worth to observe that the MathLoop is not only the fastest, but also
the most predictable, the one with the smallest standard deviation. 

String interpolation `String2Check` is not only the slowest but also the least predictable. 

## Which one to choose?

So which one to chose? The fastest always will be the fastest. But the difference is too small
to perceive in human scales, even when comparing 100,000 numbers it is at most 20ms. 

For 1,000,000 checks  the difference between the first string option and the math loop is 163ms to 12ms. 
Still almost impossible to spot with our human senses. 

And so, with performance so close, you can choose the one which is easiest to understand,

> one that reads the best. 

Reviewer: Great conclusion!

Author: Thanks
